include/openssl/hpke.h

```diff
@@ -0,0 +1,499 @@
+                             const unsigned char *aad, size_t aadlen,
+                             const unsigned char *ct, size_t ctlen);
+
+/**
+ * @brief sender export-only encapsulation function
+ * @param ctx is the pointer for the HPKE context
+ * @param enc is the sender's ephemeral public value
+ * @param enclen is the size the above
+ * @param pub is the recipient public key octets
+ * @param publen is the size the above
+ * @param info is the info parameter
+ * @param infolen is the size the above
+ * @return 1 for success, 0 for error
+ *
+ * Following this, OSSL_HPKE_export can be called.
+ */
+int OSSL_HPKE_sender_export_encap(OSSL_HPKE_CTX *ctx,
+                                  unsigned char *enc, size_t *enclen,
+                                  unsigned char *pub, size_t publen,
+                                  const unsigned char *info, size_t infolen);
+
+/**
+ * @brief recipient export-only encapsulation function
+ * @param ctx is the pointer for the HPKE context
+ * @param enc is the sender's ephemeral public value
+ * @param enclen is the size the above
+ * @param recippriv is the EVP_PKEY form of recipient private value
+ * @param info is the info parameter
+ * @param infolen is the size the above
+ * @return 1 for success, 0 for error
+ *
+ * Following this, OSSL_HPKE_export can be called.
+ */
+int OSSL_HPKE_recipient_export_decap(OSSL_HPKE_CTX *ctx,
+                                     const unsigned char *enc, size_t enclen,
+                                     EVP_PKEY *recippriv,
+                                     const unsigned char *info, size_t infolen);
+
+/**
+ * @brief generate a given-length secret based on context and label
+ * @param ctx is the HPKE context
+ * @param secret is the resulting secret that will be of length...
+ * @param secretlen is the desired output length
+ * @param label is a buffer to provide separation between secrets
+ * @param labellen is the length of the above
+ * @return 1 for good, 0 for error
+ *
+ * The context has to have been used already for one encryption
+ * or decryption for this to work (as this is based on the negotiated
+ * "exporter_secret" estabilshed via the HPKE operation).
+ */
+int OSSL_HPKE_export(OSSL_HPKE_CTX *ctx,
+                     unsigned char *secret,
+                     size_t secretlen,
+                     const unsigned char *label,
+                     size_t labellen);
+
+/*
+ * Some even newer APIs below that separate out encap/decap from
+ * seal/open/export
+ */
+
+/**
+ * @brief sender encapsulation function
+ * @param ctx is the pointer for the HPKE context
+ * @param enc is the sender's ephemeral public value
+ * @param enclen is the size the above
+ * @param pub is the recipient public key octets
+ * @param publen is the size the above
+ * @param info is the info parameter
+ * @param infolen is the size the above
+ * @return 1 for success, 0 for error
+ */
+int OSSL_HPKE_encap(OSSL_HPKE_CTX *ctx,
```

Comment:
Have you noticed that the method signatures for OSSL_HPKE_sender_export_encap() and OSSL_HPKE_encap() are the same?

Could they be the same function - if you know the mode?

Reply:
the former one will disappear imminently (see above suggest to hold for a wee bit:-)